packages/opentelemetry-plugin-document-load/src/documentLoad.ts

```diff
@@ -0,0 +1,283 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BasePlugin, otperformance } from '@opentelemetry/core';
+import { PluginConfig, Span, SpanOptions } from '@opentelemetry/types';
+import { AttributeNames } from './enums/AttributeNames';
+import { PerformanceTimingNames } from './enums/PerformanceTimingNames';
+import { PerformanceEntries, PerformanceLegacy } from './types';
+import { hasKey } from './utils';
+
+// import { IPerformanceEntries, PerformanceEntries, PerformanceLegacy } from './types';
+
+/**
+ * This class represents a document load plugin
+ */
+export class DocumentLoad extends BasePlugin<unknown> {
+  readonly component: string = 'document-load';
+  readonly version: string = '1';
+  moduleName = this.component;
+  protected _config!: PluginConfig;
+
+  /**
+   *
+   * @param {DocumentLoadConfig} config
+   */
+  constructor(config: PluginConfig = {}) {
+    super();
+    this._onDocumentLoaded = this._onDocumentLoaded.bind(this);
+    this._config = config;
+  }
+
+  /**
+   * callback to be executed when page is loaded
+   * @private
+   */
+  private _onDocumentLoaded() {
+    this._collectPerformance();
+  }
+
+  /**
+   * Collects information about performance and creates appropriate spans
+   * @private
+   */
+  private _collectPerformance() {
+    const entries = this._getEntries();
+
+    const rootSpan = this._startSpan(
+      AttributeNames.DOCUMENT_LOAD,
+      PerformanceTimingNames.FETCH_START,
+      entries
+    );
+
+    this._startAndFinishSpan(
+      AttributeNames.DOMAIN_LOOKUP,
+      PerformanceTimingNames.DOMAIN_LOOKUP_START,
+      PerformanceTimingNames.DOMAIN_LOOKUP_END,
+      entries,
+      {
+        parent: rootSpan,
+      }
+    );
+
+    // Opening Connection
+    const connectSpan = this._startSpan(
+      AttributeNames.CONNECT,
+      PerformanceTimingNames.CONNECT_START,
+      entries,
+      {
+        parent: rootSpan,
+      }
+    );
+
+    // TLS negotiation
+    if (hasKey(entries, PerformanceTimingNames.SECURE_CONNECTION_START)) {
+      const value = entries[PerformanceTimingNames.SECURE_CONNECTION_START];
+      if (typeof value === 'number' && value > 0) {
+        this._startAndFinishSpan(
+          AttributeNames.CONNECT_SECURE,
+          PerformanceTimingNames.SECURE_CONNECTION_START,
+          PerformanceTimingNames.CONNECT_END,
+          entries,
+          {
+            parent: connectSpan,
+          }
+        );
+      }
+    }
+    // Connection negotiation ends
+    this._endSpan(connectSpan, PerformanceTimingNames.CONNECT_END, entries);
+
+    // Cache seek plus response time
+    this._startAndFinishSpan(
+      AttributeNames.CACHE_SEEK,
+      PerformanceTimingNames.FETCH_START,
+      PerformanceTimingNames.RESPONSE_END,
+      entries,
+      {
+        parent: rootSpan,
+      }
+    );
+
+    // TTFB - time to first byte
```

Comment:
I like the concept of TTFB and the cache seek time, but would it make sense to have those be metrics instead of spans, and instead have the span be for the overall request time? I'm thinking that would make it more similar, particularly when joined with server side spans.